From bf4c7eea5465b1ff8d22fb94ccf04f511c64e545 Mon Sep 17 00:00:00 2001
From: Andrzej Surdej <Andrzej_Surdej@comcast.com>
Date: Fri, 6 Oct 2023 15:49:41 +0200
Subject: [PATCH] [MSE Gstreamer] Fix MSE playback pause during seek

Don't change pipeline state during ASYNC state change but apply changes
once transition is completed.
Trying to pause the playback during seek (ASYNC state change PAUSED->PAUSED)
is overwritten by PLAYING state (target state from playbin) once transition completes.
As a result pipeline is playing while MSE player thinks it's paused.
Neither HTMLVideo::pause() nor HTMLVideo::play() works
---
 .../gstreamer/mse/MediaPlayerPrivateGStreamerMSE.cpp  | 11 +++++++++++
 1 file changed, 11 insertions(+)

diff --git a/Source/WebCore/platform/graphics/gstreamer/mse/MediaPlayerPrivateGStreamerMSE.cpp b/Source/WebCore/platform/graphics/gstreamer/mse/MediaPlayerPrivateGStreamerMSE.cpp
index 61c2f376509e..4104203c01df 100644
--- a/Source/WebCore/platform/graphics/gstreamer/mse/MediaPlayerPrivateGStreamerMSE.cpp
+++ b/Source/WebCore/platform/graphics/gstreamer/mse/MediaPlayerPrivateGStreamerMSE.cpp
@@ -371,6 +371,17 @@ void MediaPlayerPrivateGStreamerMSE::updateStates()
 {
     bool shouldBePlaying = (!m_isPaused && readyState() >= MediaPlayer::ReadyState::HaveFutureData && m_playbackRatePausedState != PlaybackRatePausedState::RatePaused)
         || m_playbackRatePausedState == PlaybackRatePausedState::ShouldMoveToPlaying;
+
+    GstState state, pending;
+    GstStateChangeReturn getStateResult = gst_element_get_state(pipeline(), &state, &pending, 0);
+    if (getStateResult == GST_STATE_CHANGE_ASYNC) {
+        // Changeing pipeline state during async state change will be either rejected by changePipelineState()
+        // or ignored by playbin (overwritten by the async tartet state after transitions completes).
+        // Lets wait until it's done and then apply play/pause if needed
+        GST_DEBUG_OBJECT(pipeline(), "ASYNC state change in progress");
+        return;
+    }
+
     GST_DEBUG_OBJECT(pipeline(), "shouldBePlaying = %s, m_isPipelinePlaying = %s", boolForPrinting(shouldBePlaying), boolForPrinting(m_isPipelinePlaying));
     if (shouldBePlaying && !m_isPipelinePlaying) {
         if (!changePipelineState(GST_STATE_PLAYING))
-- 
2.25.1

